### PR TITLE
Populate Node.Status.Addresses with Hostname

### PR DESCRIPTION
**Pull request by mkulke**
This PR is supposed to address #22063 

Currently `NodeName` has to be a resolvable dns address on the master to allow apiserver -> kubelet communication (exec, log, port-forward operations on a pod). In some situations this is unfortunate (see the discussions on the issue).

The PR aims to do the following:
- Populate the `Type: Hostname` in the `Node.Status.Addresses` array, the type is already defined, but was not used so far.
- Add logic to resolve a Node's Hostname when the apiserver initiates communication with the Kubelet, instead of using the Nodename string as Hostname.

```release-note
The hostname of the node (as autodetected by the kubelet, specified via --hostname-override, or determined by the cloudprovider) is now recorded as an address of type "Hostname" in the status of the Node API object. The hostname is expected to be resolveable from the apiserver.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/25532)

### CODE REVIEW

File: pkg/kubelet/kubelet.go
Comment (liggitt) [inline pkg/kubelet/kubelet.go]: Is this hostname always resolvable from the master?

**Comment by justinsb** [comment]
ok to test

**Comment by justinsb** [comment]
Getting e2e on this asap will be helpful, I think.  I like the refactoring of the connection info into a class.  I didn't know about NodeHostName either - that's much cleaner than looking at the IPs!

File: pkg/kubelet/kubelet.go
Comment (mkulke) [inline pkg/kubelet/kubelet.go]: No, not necessarily: `// GetHostname Returns the hostname as the kubelet sees it.` Also, this can be overriden arbitrarily using `--hostname-override` on the kubelet. Maybe there are cases where this behaviour is undesired, e.g.: cloud-provider sets a resolvable Nodename, but the Node's hostname is actually not resolvable by the master, however it is preferred b/c the Addresses array is populated now. contrived?

Comment (liggitt) [inline pkg/kubelet/kubelet.go]: That's not contrived at all, imo

Comment (liggitt) [inline pkg/kubelet/kubelet.go]: I think it would be helpful to map out all the combinations (with and without cloud provider, with and without hostname override, with and without the hostname being resolvable by the master), and show when the hostname override is honored and who is responsible for setting it. I also wonder if some cloud providers need to be updated to set this info if they are known to determine hostnames that won't resolve

Comment (mkulke) [inline pkg/kubelet/kubelet.go]: @liggitt: Ok, trying to outline it:

```
$ git grep -A3 "func .* CurrentNodeName"
pkg/cloudprovider/providers/aws/aws.go:func (c *AWSCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/aws/aws.go-	return c.selfAWSInstance.nodeName, nil
pkg/cloudprovider/providers/aws/aws.go-}
pkg/cloudprovider/providers/aws/aws.go-
--
pkg/cloudprovider/providers/fake/fake.go:func (f *FakeCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/fake/fake.go-	return hostname, nil
pkg/cloudprovider/providers/fake/fake.go-}
pkg/cloudprovider/providers/fake/fake.go-
--
pkg/cloudprovider/providers/gce/gce.go:func (gce *GCECloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/gce/gce.go-	return hostname, nil
pkg/cloudprovider/providers/gce/gce.go-}
pkg/cloudprovider/providers/gce/gce.go-
--
pkg/cloudprovider/providers/mesos/mesos.go:func (c *MesosCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/mesos/mesos.go-	return hostname, nil
pkg/cloudprovider/providers/mesos/mesos.go-}
pkg/cloudprovider/providers/mesos/mesos.go-
--
pkg/cloudprovider/providers/openstack/openstack.go:func (i *Instances) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/openstack/openstack.go-	return hostname, nil
pkg/cloudprovider/providers/openstack/openstack.go-}
pkg/cloudprovider/providers/openstack/openstack.go-
--
pkg/cloudprovider/providers/ovirt/ovirt.go:func (v *OVirtCloud) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/ovirt/ovirt.go-	return hostname, nil
pkg/cloudprovider/providers/ovirt/ovirt.go-}
pkg/cloudprovider/providers/ovirt/ovirt.go-
--
pkg/cloudprovider/providers/rackspace/rackspace.go:func (i *Instances) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/rackspace/rackspace.go-	// Beware when changing this, nodename == hostname assumption is crucial to
pkg/cloudprovider/providers/rackspace/rackspace.go-	// apiserver => kubelet communication.
pkg/cloudprovider/providers/rackspace/rackspace.go-	return hostname, nil
--
pkg/cloudprovider/providers/vsphere/vsphere.go:func (i *Instances) CurrentNodeName(hostname string) (string, error) {
pkg/cloudprovider/providers/vsphere/vsphere.go-	return i.localInstanceID, nil
pkg/cloudprovider/providers/vsphere/vsphere.go-}
pkg/cloudprovider/providers/vsphere/vsphere.go-
```

Most map directly to hostname, VSphere and AWS don't. 

* VSphere maps to VM name, not necessarily resolvable. Right now the VM name has to be configured to be a master-resolvable DNS name.

https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/vsphere/vsphere.go#L127

* AWS gets its hostname from the API, using `instance.PrivateDnsName`, which is equal to an EC2 instance hostname:

  https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L982

The current kubelet code will get a Node's hostname by executing `uname -n`, unless `--hostname-override` is set, then it will use this value.

In general I don't think this is wrong (actually it's the intended behavior IMO): It's better to require the Node's hostname to be master-resolvable than have a convention that the Nodename has to be resolvable.

However this could introduce a breaking change for clusters set up under the assumption: Nodename matters, hostname doesn't. To make sure this doesn't happen we can add the `Type:Hostname` Address to the address array in the affected cloud-providers code already, and add it in the kubelet code only when it's not set by the cloud-provider (actually this sanity-check should be there anyway).

**Comment by justinsb** [comment]
This is a fairly significant change (operationally, not in LOC).  I propose we try to land it in 1.4.

@mkulke you'll probably have to ping me once 1.4 opens to remind me..

**Comment by smarterclayton** [comment]
I agree with that - the risk is very high.

File: pkg/registry/pod/strategy.go
Comment (justinsb) [inline pkg/registry/pod/strategy.go]: Aha!  So the thing is that here connInfo is a REST object, not a "raw" HTTPKubeletClient object.  Because the one will convert the the nodeName and the other will not.  I'm going to verify that; but I propose changing REST to implement a different interface, given the behaviours are different (convert node name vs not).

**Comment by justinsb** [comment]
I think this is great, and exactly the sort of thing that has made turn-up hard.  (cc @mikedanese )

It is high risk, but we are at the stage where we can implement it I think.

But I think we should just the same logic as we have in 2d85e4a0 in the pod provider.  Arguably then this is actually a bug fix, in that it is inconsistent that the pods package resolve the kubelet differently from how the nodes packages resolves it.  That is arguable, but the concrete consequence would be that we should not have to change the kubelet/node registration at all, and that we should prefer the InternalIP.  That is what InternalIP is for, I believe.

If there is a provider for whom this is unworkable, we have plenty of time to discover it, and even add yet-another-address-type if that is needed.

File: pkg/registry/node/etcd/etcd.go
Comment (liggitt) [inline pkg/registry/node/etcd/etcd.go]: this is forcing two lookups of the node object from etcd... (once for the host, once for the port)... just fetch once

File: pkg/registry/pod/strategy.go
Comment (liggitt) [inline pkg/registry/pod/strategy.go]: I don't think it needs a different interface... the inputs are a nodeName, the outputs are the connection information. In the absence of this new address info on a node, the fallback will be to continue using the node name as the host.

**Comment by justinsb** [comment]
I just put #29011 which is related.  This PR assumes that we always call GetConnectionInfo through the etcd REST class implementation.  This is not obvious though, because KubeletClient implements GetConnectionInfo.  #29011 is a small refactor to ensure that GetConnectionInfo is not implemented by KubeletClient, ensuring that the only real implementation of GetConnectionInfo is the etcd REST class here, which means we are guaranteed to go through this logic.  (I think :-) )

File: pkg/registry/node/etcd/etcd.go
Comment (mkulke) [inline pkg/registry/node/etcd/etcd.go]: @liggitt: make sense, i'm retrieving both in a single query now.

**Comment by liggitt** [comment]
> we should prefer the InternalIP

That will cause TLS problems. DNS is preferred.

**Comment by justinsb** [comment]
> Are you saying we should prefer the internal IP for master->node communications?

Exactly right - I think that's what this PR is about... currently we assume the node name is resolvable from the master to reach kubelet on the node.  I think we should use the internal IP.

If this then fails TLS hostname validation checks because of the way we build certs, maybe we'll then have to pass specify both the hostname (certname) & the IP when we're connecting master -> node.

**Comment by liggitt** [comment]
I thought this PR was about populating a hostname address that could resolve, not switching to use internal IP addresses

**Comment by justinsb** [comment]
@liggit the problem we're trying to solve is that the nodename must currently be a resolvable name on the master.  That was a GCE assumption that doesn't really make sense on other platforms, and I for one have been trying for a long time to remove the link.  I think it also makes bringup much harder - certainly a lot of people have trouble with it.

I guess I am suggesting an alternate way of achieving that to the way this PR currently does it, but I think it's simpler.  More importantly, it is consistent with https://github.com/kubernetes/kubernetes/commit/2d85e4a094b1f3099dce6ac9ad540a46e54184be 

Looking at that logic, we connect to the node by internal IP in the proxy case.  I think switching to use that rule everywhere would therefore break no functioning cluster, and would help a lot of people build working clusters without trying to sync their hostnames with their nodenames.  For example, on AWS, it might make more sense to refer to nodes by their AWS instance ids, or by the correct hostname when a custom DHCP subdomain is set.

Unless you're going to tell me we have yet another piece of node-resolver logic that uses the Hostname Node address somewhere ... :-(

**Comment by liggitt** [comment]
I understand the problem, but DNS is preferred in some environments.

> If this then fails TLS hostname validation checks because of the way we build certs, maybe we'll then have to pass specify both the hostname (certname) & the IP when we're connecting master -> node.

I am not in favor of non-standard TLS validation, or of defaulting to using an IP, which can cause problems getting serving certs from some CA.

> More importantly, it is consistent with 2d85e4a ... Looking at that logic, we connect to the node by internal IP in the proxy case. I think switching to use that rule everywhere would therefore break no functioning cluster
>
> Unless you're going to tell me we have yet another piece of node-resolver logic that uses the Hostname Node address somewhere ... :-(

Yes, the majority of connection methods use DNS, using the node name: attaching/execing into pods, port forwarding to pods, streaming logs from pods

It is unfortunate that 2d85e4a is not consistent, but node proxy use is *much* less common than all the pod connection methods (which use DNS), so I am *more* confident the node-name-as-DNS method is working in most currently running clusters.

I am in favor of the hostname change, in order to break the requirement that the node *name* be resolveable (since it is immutable), but I don't think we should switch to internal IP by default.

File: pkg/registry/node/etcd/etcd.go
Comment (mkulke) [inline pkg/registry/node/etcd/etcd.go]: @liggitt: regarding the noisiness of this log message: the occurrence should be rare. `Hostname` should be set always (https://github.com/kubernetes/kubernetes/pull/25532/files#diff-bf28da68f62a8df6e99e447c4351122dR2718), it's just a defensive measure (apiserver/kubelet version mismatch maybe)

**Comment by mkulke** [comment]
@justinsb @liggitt to chime in on the aim of the PR: the idea was indeed to decouple a Node's `NodeName` from how the Master is addressing that Node (since the assumptions that come with coupling both did not work for us and others in certain situations, see issue linked above). 

I'm not particularly opinionated about a specific default, there are good reasons for using domain names over IPs, as @liggitt mentioned. we should provide means to override that default, though (by specifying `--override-hostname` i'd say).

File: pkg/kubelet/kubelet_node_status.go
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: why not?

File: pkg/registry/node/etcd/etcd.go
Comment (liggitt) [inline pkg/registry/node/etcd/etcd.go]: when there's a mismatch, how noisy is this? that occurs for some period of time during an upgrade

File: pkg/registry/node/etcd/etcd.go
Comment (liggitt) [inline pkg/registry/node/etcd/etcd.go]: this is odd... we resolve `node -> host,port`, then resolve `hostname -> {scheme,host,port,transport}`, then ignore the second port and replace it with the first?

Comment (liggitt) [inline pkg/registry/node/etcd/etcd.go]: also, `GetConnectionInfo` takes a nodeName, not a hostname

Comment (liggitt) [inline pkg/registry/node/etcd/etcd.go]: I'm remembering this now... `c.connection` is the default connection info, not the per-node info lookup

File: pkg/registry/pod/strategy.go
Comment (liggitt) [inline pkg/registry/pod/strategy.go]: while you're in here, `net.JoinHostPort(connectionInfo.Hostname, connectionInfo.Port)`

File: pkg/registry/pod/strategy.go
Comment (liggitt) [inline pkg/registry/pod/strategy.go]: `net.JoinHostPort(connectionInfo.Hostname, connectionInfo.Port)`

File: pkg/registry/node/etcd/etcd.go
Comment (liggitt) [inline pkg/registry/node/etcd/etcd.go]: this might read a little better if we swapped the order:

```
// compute the generic connection info
connectionInfo, err := c.connection.GetConnectionInfo(ctx, nodeName)
if err != nil {
  return nil, err
}

// get any specific host/port info for this node
hostname, port, err := c.getKubeletHostPortPair(nodeName)
if err != nil {
  return nil, err
}
if len(hostname) > 0 {
  connectionInfo.Host = hostname
}
if port > 0 {
  connectionInfo.Port = port
}

return connectionInfo, nil
```

File: pkg/kubelet/client/kubelet_client.go
Comment (liggitt) [inline pkg/kubelet/client/kubelet_client.go]: use field names in initialization (e.g. `&ConnectionInfo{Scheme:scheme, Hostname: nodeName, ...}`

File: pkg/registry/node/etcd/etcd.go
Comment (mkulke) [inline pkg/registry/node/etcd/etcd.go]: @liggitt: basically every time a user is initiating apiserver -> kubelet comm, i.e. when retrieving logs, on exec, on port-forward.

File: pkg/kubelet/kubelet_node_status.go
Comment (mkulke) [inline pkg/kubelet/kubelet_node_status.go]: @liggitt The rationale behind it is: currently the cloud providers do not use do this, we can assume a node to know its hostname, and users specifying a hostname via `--hostname-override` on kubelet invocation do not want the cloud provider to have override this w/ their value (we do not trace whether the user did a hostname override, we consider it initially in `util/node/node.go`).

Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: it seems like the cloud provider would be best positioned to know the hostname in some cases (e.g. the AWS provider querying "public-hostname" from metadata). I agree the interaction with user-specified --hostname-override is weird, I just don't know that the cloudprovider should be forbidden from resolving a hostname

File: pkg/registry/pod/strategy.go
Comment (mkulke) [inline pkg/registry/pod/strategy.go]: @liggitt `connectionInfo.Port` is `uint`, while `JoinHostPort` wants a string, so sprintfs conversion would be required anyway, should we still use it?

Comment (liggitt) [inline pkg/registry/pod/strategy.go]: if the hostname is ever an IPv6 IP, the sprintf will be wrong, so... probably?

Comment (liggitt) [inline pkg/registry/pod/strategy.go]: fine to do separately, since the issue is preexisting

File: pkg/kubelet/kubelet_node_status.go
Comment (mkulke) [inline pkg/kubelet/kubelet_node_status.go]: @liggitt hmm true, it's probably a bit arbitrary. We could defer handling the hostname-override conflict and remove that restriction until the issue actually pops up (i.e. a cloud provider setting the `NodeHostName` prop and ignoring `--hostname-override`). I suspect the proper way to handle it would be to persist the hostname-override intent somewhere, but this can be dealt with in another PR.

**Comment by luxas** [comment]
ping

Will this make v1.4? @pwittrock @justinsb?

Would be great to have...

**Comment by liggitt** [comment]
definitely not, 1.4 is frozen except for blocking bugs, and this is changing behavior that has existed since 1.0

**Comment by luxas** [comment]
Okay, that was what I supposed; it's a feature, not a bug

bare-metal guys have to workaround it with `--hostname-override` until v1.5 then...

**Comment by justinsb** [comment]
If we're going to stick with Type=HostName I believe we need a feature & design doc.

**Comment by mkulke** [comment]
@justinsb are you sure? `Type: Hostname` has been there for quite some time, albeit unused.

**Comment by justinsb** [comment]
@mkulke - not for populating the Addresses, but for using it for resolution, yes.  We currently have two behaviours: using the IP addresses in the order specified by nodeutil.GetNodeHostIP, and using the NodeName.

My understanding is that NodeName resolution is just a bug - we should be using nodeutil.GetNodeHostIP.

I think adding resolution based on HostName is a 3rd mechanism and a (rather big) behaviour change, and thus needs some form of justification.

If HostName was just considered as a fallback after GetNodeHostIP I could see this as straight PR, but otherwise it's a pretty big change if it takes precedence.

**Comment by liggitt** [comment]
> My understanding is that NodeName resolution is just a bug - we should be using nodeutil.GetNodeHostIP.

I wouldn't characterize it as a bug, and preferring the node IP for master->node communication makes TLS more difficult. Letting the node report the host name to use rather than assuming the node name is DNS resolveable seems like the right approach to me.

**Comment by justinsb** [comment]
@liggitt certificate validation is a separate issue - this is about connecting to the node.  I think you can't use hostname for certificate validation anyway: https://github.com/kubernetes/kubernetes/blob/release-1.4/cmd/kubelet/app/bootstrap.go#L194-L206

I wonder if we should store the certificate public key (fingerprint) in the node object, or maybe even have a strongly typed certificate validation policy.  I think breaking the link between connecting and certificate validation is both necessary (for better or worse) but also opens up some exciting possibilities for stronger validation than pure hostname validation.

**Comment by liggitt** [comment]
> @liggitt certificate validation is a separate issue - this is about connecting to the node

how you connect to the kubelet API is intimately related to certificate validation of the kubelet's serving certificates. What you linked to is the CSR for the client certificate the kubelet uses to talk back to the master, and is unrelated to the node's serving certs.

> I wonder if we should store the certificate public key (fingerprint) in the node object, or maybe even have a strongly typed certificate validation policy. I think breaking the link between connecting and certificate validation is both necessary (for better or worse) but also opens up some exciting possibilities for stronger validation than pure hostname validation.

I am strongly opposed to non-standard validation schemes... more things talk to the kubelet API than just the master. Depending on non-standard cert validation schemes almost guarantee that TLS validation will simply be disabled.

**Comment by justinsb** [comment]
> how you connect to the kubelet API is intimately related to certificate validation of the kubelet's serving certificates. What you linked to is the CSR for the client certificate the kubelet uses to talk back to the master, and is unrelated to the node's serving certs.

Ah thanks for the correction.  Where is the code I should be looking at?

**Comment by liggitt** [comment]
> Ah thanks for the correction. Where is the code I should be looking at?

Not sure what you are looking for. The kubelet takes `--tls-cert-file` and `--tls-private-key-file` args for the serving cert/key to use. The apiserver takes a `--kubelet-certificate-authority` arg for the CA bundle to use to verify the kubelet's serving cert. A properly secured deployment would set both of those.

**Comment by mkulke** [comment]
> I wouldn't characterize it as a bug, and preferring the node IP for master->node communication makes TLS more difficult. Letting the node report the host name to use rather than assuming the node name is DNS resolveable seems like the right approach to me.

I agree here, using IPs for master->node communication should not be the default option, but it should ultimately be available as an option via `--hostname-override`, when hostname discovery by the master is not feasible for some reason.

**Comment by bgrant0607** [comment]
cc @cjcullen

**Comment by liggitt** [comment]
let's limit this to recording the detected (or `--override-hostname`) hostname as a Hostname address.

I opened https://github.com/kubernetes/kubernetes/pull/34474 as a follow up to https://github.com/kubernetes/kubernetes/pull/33718 and added the ConnectionInfo struct there, and will add address-type preference ordering in a separate PR

**Comment by mkulke** [comment]
@liggitt ok, i'll wait until #33718 is merged and then rebase.

**Comment by liggitt** [comment]
#33718 and #34474 have merged if you'd like to go ahead and rebase this to just record the hostname address

File: pkg/kubelet/kubelet_node_status.go
Comment (liggitt) [inline pkg/kubelet/kubelet_node_status.go]: inline `api.NodeAddress{Type: api.NodeHostName, Address: kl.GetHostname()}` to make it clear what is being returned

**Comment by liggitt** [comment]
This PR records the kubelet hostname (autodetected or manually specified with `--hostname-override`) in Node status as an address of type `Hostname`.

Tagging @kubernetes/api-review-team for approval. Previous discussion at https://github.com/kubernetes/kubernetes/pull/33718#issuecomment-251864116.

This PR and https://github.com/kubernetes/kubernetes/pull/35497 are needed by 1.5 to preserve the ability to use a DNS name for master->node communications (typically for SSL cert purposes)

**Review by liggitt — CHANGES_REQUESTED** [review]
Include the hostname address in the nodeIP-filtered list, and needs sign-off from @kubernetes/api-review-team / @kubernetes/kube-api , LGTM otherwise

**Comment by liggitt** [comment]
#33388 flake
@k8s-bot gci gke e2e test this

**Comment by liggitt** [comment]
@mkulke, can you make the last couple changes, would like to get this merged early this week ahead of code-freeze

**Comment by smarterclayton** [comment]
This looks reasonable to me - breaking backwards compatibility with node name lookups in existing clusters is deeply dangerous, so this appears to correctly preserve legacy behavior.

**Comment by mkulke** [comment]
@liggitt: ok, i addressed the review comments, squashed and rebased.

**Comment by liggitt** [comment]
#35898 flake
@k8s-bot node e2e test this
@k8s-bot unit test this

**Review by liggitt — APPROVED** [review]

**Comment by liggitt** [comment]
LGTM. Needs a nod from @kubernetes/api-review-team for the use of the `Hostname` node address type (previous discussion at https://github.com/kubernetes/kubernetes/pull/33718#issuecomment-251864116)

cc @bgrant0607 @cjcullen

**Comment by bgrant0607** [comment]
ref #9267

**Comment by bgrant0607** [comment]
@liggitt I'm fine with this, and happy if it gets us closer to moving away from resolving node resource names. Please document the required level of resolvability of the hostname (e.g., must be resolvable by the apiserver).

However, I haven't thought about what more we could do to determine whether it might break anyone, or all the different ways it might interact with Kubelet configuration.

At minimum, this needs a much more detailed release note than just the title of this PR.

Something to think about for the future: At some point we discussed adding info about which form of the address to use to verify node identity.

**Comment by bgrant0607** [comment]
cc @caesarxuchao

**Comment by liggitt** [comment]
Updated release-note with description of source and expectations for the address. @mkulke can you open a doc PR as well?

**Pull request merged**
